eth2near/utilities/src/eth_proof_generator.rs

```diff
@@ -0,0 +1,178 @@
+
+pub fn get_proof_for_event(tx_hash: &str, log_index: u64, node_url: &str,) -> Result<Proof, Box<dyn Error>> {
+    let client = EthRPCClient::new(node_url);
+
+    let tx_bytes: [u8; 32] = hex::decode(&tx_hash[2..])
+        .expect("Invalid hex string")
+        .try_into()
+        .expect("Invalid hex string");
+
+    let receipt = client.get_transaction_receipt_by_hash(&FixedBytes(tx_bytes))?;
+    let block_header = client.get_block_by_number(receipt.block_number)?;
+    let block_receipts = client.get_block_receipts(receipt.block_number)?;
+
+    let trie = build_receipt_trie(&block_receipts)?;
+
+    let receipt_key = rlp::encode(&receipt.transaction_index);
+    let proof = trie.get_proof(&receipt_key)?;
+
+    let mut log_data: Option<Vec<u8>> = None;
+    for (_, log) in receipt.logs.iter().enumerate() {
+        if log.log_index == log_index {
+            log_data = Some(encode_log(log));
+        }
+    }
+
+    Ok(Proof {
+        log_index,
+        log_entry_data: log_data.expect("Log not found"),
+        receipt_index: receipt.transaction_index,
+        receipt_data: encode_receipt(&receipt),
+        header_data: encode_header(&block_header),
+        proof,
+    })
+}
+
+fn build_receipt_trie(receipts: &[TransactionReceipt],) -> Result<PatriciaTrie<MemoryDB, HasherKeccak>, TrieError> {
+    let memdb = Arc::new(MemoryDB::new(true));
+    let hasher = Arc::new(HasherKeccak::new());
+    let mut trie = PatriciaTrie::new(Arc::clone(&memdb), Arc::clone(&hasher));
+
+    for (_, receipt) in receipts.iter().enumerate() {
+        let receipt_key = rlp::encode(&receipt.transaction_index).to_vec();
+        let receipt_data = encode_receipt(receipt);
+
+        trie.insert(receipt_key, receipt_data)?;
+    }
+
+    Ok(trie)
+}
+
+fn encode_receipt(receipt: &TransactionReceipt) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+
+    if receipt.transaction_type != 0 {
+        stream.append(&receipt.transaction_type);
+    }
+
+    stream.begin_list(4);
+    stream
+        .append(&receipt.status)
+        .append(&receipt.cumulative_gas_used.0)
+        .append(&receipt.logs_bloom.0.to_vec());
+
+    stream.begin_list(receipt.logs.len());
+    for (_, log) in receipt.logs.iter().enumerate() {
+        stream.begin_list(3);
+        stream.append(&log.address.0.to_vec());
+
+        stream.begin_list(log.topics.len());
+        for (_, topic) in log.topics.iter().enumerate() {
+            stream.append(&topic.0.to_vec());
+        }
+
+        stream.append(&log.data.0);
+    }
+
+    stream.out().to_vec()
+}
+
+fn encode_log(log: &Log) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+    stream.begin_list(3);
+
+    stream.append(&log.address.0.to_vec());
+
+    stream.begin_list(log.topics.len());
+    for (_, topic) in log.topics.iter().enumerate() {
+        stream.append(&topic.0.to_vec());
+    }
+
+    stream.append(&log.data.0);
+
+    stream.out().to_vec()
+}
+
+fn encode_header(header: &BlockHeader) -> Vec<u8> {
```

Comment:
I think it's because eth-types are under contracts namespace so we avoid using it outside of NEAR smart contracts. Traits could be a cleaner approach I agree